### ISSUE TITLE
Unnecessary parentheses warning for (A | B) as :pat in 2018 edition

### ISSUE THREAD

**Issue by ExpHP** [rust-lang/rust#86959]
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=30f392098326a970e468c3125f7c5413
```rust
macro_rules! mac {
    ($expr:expr, $pat:pat) => {
        match $expr {
            $pat => {},
            _ => panic!(),
        }
    };
}

fn main() {
    mac!('a', ('a' | 'A'));
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
warning: unnecessary parentheses around pattern
  --> src/main.rs:12:15
   |
12 |     mac!('a', ('a' | 'A'));
   |               ^^^^^^^^^^^ help: remove these parentheses
   |
   = note: `#[warn(unused_parens)]` on by default

warning: 1 warning emitted
```

While it would be true that these parentheses are unnecessary in 2021 edition, they are required in 2018 edition because `:pat` cannot match `'a' | 'A'` in this edition, making the warning a nuisance.

**Comment by apiraino** [comment]
I've tried bisecting this and it seems that the warning was introduced a long time ago, exactly on `nightly-2019-08-19` (before was a hard error). Probably goes back to https://github.com/rust-lang/rust/issues/54883

```
found 7 bors merge commits in the specified range
  commit[0] 2019-08-17UTC: Auto merge of #63658 - RalfJung:miri-op, r=oli-obk
  commit[1] 2019-08-17UTC: Auto merge of #63671 - Centril:rollup-zufavt5, r=Centril
  commit[2] 2019-08-18UTC: Auto merge of #61708 - dlrobertson:or-patterns-0, r=centril
  commit[3] 2019-08-18UTC: Auto merge of #62948 - matklad:failable-file-loading, r=petrochenkov
  commit[4] 2019-08-18UTC: Auto merge of #63269 - Aaron1011:feature/proc-macro-data, r=eddyb,petrochenkov
  commit[5] 2019-08-18UTC: Auto merge of #63635 - oli-obk:default-slice-dangles, r=eddyb
  commit[6] 2019-08-18UTC: Auto merge of #63659 - gilescope:async-in-closure, r=Centril
ERROR: no commits between 2111aed0a38c819acb140c7153e9366964a37f2f and 4cf7673076e6975532213e494dd3f7f9d8c2328e within last 167 days
```

bisectecd with
```
RUST_SRC_REPO=~/projects/rust/rust cargo-bisect-rustc --preserve --start 2019-01-01 --script=./test.sh
```

and this helper script
```
#!/usr/bin/env bash
out=$(cargo check 2>&1)
if echo "${out}" | grep -q "remove these paren"; then
    exit 1
fi
exit 0
```

**Comment by ExpHP** [comment]
Yes, I wouldn't expect bisecting to tell much in this instance because I wouldn't expect this to be a regression?  I could be wrong, but it seems pretty clear how this behavior arises as a logical consequence of other behaviors of the compiler, and normally "unused" warnings are just a basic fact of life when working with macros.

I created an issue for this warning because I feel that this one is particularly easy to run into; there is a clear use case for writing parenthesized `|` patterns in 2018 edition, and those parens will almost always be "unused" once expanded.

**Comment by apiraino** [comment]
@ExpHP yes, totally agree. I've tried a diagnose because I was confused about the "edition 2021" part of your report: also trying to make this code "2021-ready", the warning does not disappear without forcing it out by adding `#![allow(unused_parens)]` (which is a bit inconvenient).

Perhaps related to #80636 and (in cascade) to #78747

Assigning priority as discussed in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2386959.20Unnecessary.20parentheses.20warning.20for.20.28A.20.7C.20B.29.20as.20.3Apat.20.E2.80.A6) of the Prioritization Working Group.

@rustbot label -I-prioritize +P-mdium
